Scrapper-Service/plugins/guide2research.py

```diff
@@ -0,0 +1,378 @@
+import logging
+from datetime import datetime as dt
+from typing import Dict, List
+import itertools
+import requests
+from bs4 import BeautifulSoup as bs
+
+from commons import PageParsingError, Scrapper
+from datamodels import Conference, Metadata
+
+
+class Guide2ResearchScrapper(Scrapper):
+    def __init__(self, **config):
+        super().__init__(context_name=__name__, **config)
+        self.base_address = "http://www.guide2research.com/"
+        self.top_conf_address = self.base_address + "topconf/"
+        self.all_confs_base_address = self.base_address + "conferences/"
+        self.all_conf_page_links = [self.all_confs_base_address]
+        self.all_conf_page_count = 1
+        self.site_name = __name__
+        self.scrapper_name = "Guide2Research"
+
+    def parse_action(self):
+        self.extract_and_push()
+
+    def extract_and_push(self):
+        """
+        Main calling method for g2r scraper
+        Collects the links of all top conferences available
+        and posts them to db
+        """
+        top_conference_links = self._get_conferences_list(which='top')
+        all_conference_links = self._get_conferences_list(which='all')
+        for i, link in enumerate(top_conference_links):
+            try:
+                conf_data = self._parse_top_conference(link=link)
+                self.push_todb(conf_data)
+                self.logger.info(f"Number of Top Conferences added: {i} ")
+            except Exception as e:
+                self.logger.error(
+                    f"Error while parsing page {link}, find full trace {e}")
+        for i, link in enumerate(all_conference_links):
+            try:
+                conf_data = self._parse_all_conference(link=link)
+                self.push_todb(conf_data)
+                self.logger.info(f"Number of All Conferences added: {i}")
+            except Exception as e:
+                self.logger.error(
+                    f"Error while parsing {link}, full trace {e}"
+                )
+
+    def _get_conferences_list(self, which: str) -> List[str]:
+        """
+        parses topconferences baseurl and returns -
+        - a list of available links
+        """
+        if which == 'all':
+            self._get_all_conf_pages(url=self.all_confs_base_address)
+            links = []
+            for link in self.all_conf_page_links:
+                content = self.get_page(
+                    qlink=link, debug_msg=f"Extracted links from {link}").content
+                links.append(self._parse_all_conference_base(content=content))
```

Comment:
Can you suggest a way of going about it by looking at the flow?
Here's an outline
1. First, I'm looking for all pages of all conferences and saving them to `all_conf_page_links`
2. Then I'm further parsing each of the item in that list to get links individual conferences and adding them to a list.

Reply:
Please look at lines

Reply:
Follow the below patter
```python
 def some_function_that_return_list( args ):
            while True:
                 ## do some processing 
                 yield result

 def other_func_that_uses_above( args ):
       for result in some_function_that_return_list(new_args):
             ## do something with result
             new_result = process(result)
             yield result

 for result in other_funct_that_uses_above(args):
      print(result)
```

Where you see your self iterating over a one-timer list, you can resort to yield. like `self.all_conf_page_links`  , can be a yielded instead of a long list. Similarly the processing done on it can be yield also ........ The idea is dont accumalate into a list just to return it. Yield the list..